packages/abi/src/coder/encoding/v1/enum.ts

```diff
@@ -0,0 +1,90 @@
+import { concat } from '@fuel-ts/utils';
+import type { RequireExactlyOne } from 'type-fest';
+
+import { ENUM_TYPE, VOID_TYPE } from '../encoding-constants';
+import type { Coder, TypesOfCoder } from '../encoding-types';
+import { errors } from '../validation';
+
+import { u64 } from './fixed';
+
+export type EnumEncodeValue<TCoders extends Record<string, Coder>> = RequireExactlyOne<{
+  [P in keyof TCoders]: TypesOfCoder<TCoders[P]>['Input'];
+}>;
+export type EnumDecodeValue<TCoders extends Record<string, Coder>> = RequireExactlyOne<{
+  [P in keyof TCoders]: TypesOfCoder<TCoders[P]>['Decoded'];
+}>;
+export type EnumCoder<TCoders extends Record<string, Coder> = Record<string, Coder>> = Coder<
+  EnumEncodeValue<TCoders>,
+  EnumDecodeValue<TCoders>
+>;
+
+const isNativeCoder = (coder: Coder) => coder.type === VOID_TYPE;
+const extractCaseKeysFromValue = <TCoders extends Record<string, Coder>>(
+  value: EnumEncodeValue<TCoders>
+): string[] => (typeof value === 'string' ? [value] : Object.keys(value));
+const extractCaseValueFromValue = <TCoders extends Record<string, Coder>>(
+  value: EnumEncodeValue<TCoders>
+) => (typeof value === 'string' ? undefined : Object.values(value)[0]);
+const extractCaseIndexFromData = (data: Uint8Array, initialOffset: number) => {
+  const [decoded, offset] = u64.decode(data, initialOffset);
+  return [decoded.toNumber(), offset];
```

Comment:
Just double checking here, but given JS's [MAX_SAFE_INTEGER](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/MAX_SAFE_INTEGER) is 2^53 is there a potential for loss of precision for a `u64`  by using `toNumber()` ?

Reply:
If the number is too big this can throw as well, I think

Reply:
You are correct; however, that would imply you're working with an object with more than 2^53 keys.

Reply:
> You are correct; however, that would imply you're working with an object with more than 2^53 keys.

I created an issue in the spec to change this: https://github.com/FuelLabs/fuel-specs/issues/622.

Reply:
Thanks for clarifying, I'll resolve this then.